### PR TITLE
Fix hardcoded conditional rendering of client secret input field

### LINKED ISSUE #22660: Implementing custom ClientAuthenticator loses access to Client Secret Input Field in the Admin UI
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the [latest nightly release](https://github.com/keycloak/keycloak/releases/tag/nightly)

### Area

admin/ui

### Describe the bug

In the client credentials tab, the `ClientSecret` input field in the UI is conditionally rendered with a hardcoded names of ClientAuthenticator implementation. [Link to code line](https://github.com/keycloak/keycloak/blob/660d17f8ec113558e520be97007f6b4a72f0fca6/js/apps/admin-ui/src/clients/credentials/Credentials.tsx#L208C28-L208C28)

```jsx
{(clientAuthenticatorType === "client-secret" ||
            clientAuthenticatorType === "client-secret-jwt") && (
    <CardBody>
        <ClientSecret
            client={client}
            secret={secret}
            toggle={toggleClientSecretConfirm}
         />
    </CardBody>
)}
```

### Version

21.1.1

### Expected behavior

Client Secret conditional rendering should depend on the value of

```java
boolean supportsSecret() {
        return true;
 }
```

which is defined by the interface [ClientAuthenticatorFactory](https://github.com/keycloak/keycloak/blob/660d17f8ec113558e520be97007f6b4a72f0fca6/server-spi-private/src/main/java/org/keycloak/authentication/ClientAuthenticatorFactory.java#L37)

This value can be brought to frontend using the existing REST endpoint that keycloak Admin UI uses to get a list of client authenticators. `/admin/realms/${realm-name}/authentication/client-authenticator-providers`

The current response of this endpoint is as follows

```json
[
    {
        "displayName": "Signed Jwt",
        "description": "Validates client based on signed JWT issued by client and signed with the Client private key",
        "id": "client-jwt"
    },
    {
        "displayName": "Client Id and Secret with license check.",
        "description": "Validates client based on 'client_id' and 'client_secret' sent either in request parameters or in 'Authorization: Basic' header. It also ensures the realm have a valid license.",
        "id": "client-secret-license"
    },
    {
        "displayName": "Client Id and Secret",
        "description": "Validates client based on 'client_id' and 'client_secret' sent either in request parameters or in 'Authorization: Basic' header",
        "id": "client-secret"
    },
    {
        "displayName": "X509 Certificate",
        "description": "Validates client based on a X509 Certificate",
        "id": "client-x509"
    },
    {
        "displayName": "Signed Jwt with Client Secret",
        "description": "Validates client based on signed JWT issued by client and signed with the Client Secret",
        "id": "client-secret-jwt"
    }
]
```

The response of this endpoint can be extended to include the the `supportsSecret` attribute as follows:

```json
[
    {
        "displayName": "Signed Jwt",
        "description": "Validates client based on signed JWT issued by client and signed with the Client private key",
        "id": "client-jwt",
        "supportsSecret": false
    },
    {
        "displayName": "Client Id and Secret with license check.",
        "description": "Validates client based on 'client_id' and 'client_secret' sent either in request parameters or in 'Authorization: Basic' header. It also ensures the realm have a valid license.",
        "id": "client-secret-license",
        "supportsSecret": true
    },
    {
        "displayName": "Client Id and Secret",
        "description": "Validates client based on 'client_id' and 'client_secret' sent either in request parameters or in 'Authorization: Basic' header",
        "id": "client-secret",
        "supportsSecret": true
    },
    {
        "displayName": "X509 Certificate",
        "description": "Validates client based on a X509 Certificate",
        "id": "client-x509",
        "supportsSecret": false
    },
    {
        "displayName": "Signed Jwt with Client Secret",
        "description": "Validates client based on signed JWT issued by client and signed with the Client Secret",
        "id": "client-secret-jwt",
        "supportsSecret": true
    }
]
```

Which can be done in [AuthenticationManagementResource](https://github.com/keycloak/keycloak/blob/660d17f8ec113558e520be97007f6b4a72f0fca6/services/src/main/java/org/keycloak/services/resources/admin/AuthenticationManagementResource.java#L164) as follows:

```java
public Stream<Map<String, Object>> buildProviderMetadata(Stream<ProviderFactory> factories) {
        return factories.map(factory -> {
            Map<String, Object> data = new HashMap<>();
            data.put("id", factory.getId());
            ClientAuthenticatorFactory configured = (ClientAuthenticatorFactory)factory;
            data.put("description", configured.getHelpText());
            data.put("displayName", configured.getDisplayType());
            data.put("supportsSecret", configured.supportsSecret());
            return data;
       });
}
```

In the Admin UI source code, the conditional rendering of the client secret field can then depends of this value

```jsx
{(clientAuthenticatorType.supportsSecret == true) && (
    <CardBody>
        <ClientSecret
            client={client}
            secret={secret}
            toggle={toggleClientSecretConfirm}
         />
    </CardBody>
)}
```

### Actual behavior

client secret input field, will only be rendered when the PROVIDER_ID is`client-secret` or `client-secret-jwt`.

When attempted to implement ClientAuthenticator, The access to the client secret field is lost, because the provider name cannot be  `client-secret` or `client-secret-jwt`, as they are already in use.

### How to Reproduce?

Implement [ClientAuthenticator](https://github.com/keycloak/keycloak/blob/660d17f8ec113558e520be97007f6b4a72f0fca6/server-spi-private/src/main/java/org/keycloak/authentication/ClientAuthenticator.java#L34C11-L34C11) or extend [AbstractClientAuthenticator](https://github.com/keycloak/keycloak/blob/660d17f8ec113558e520be97007f6b4a72f0fca6/services/src/main/java/org/keycloak/authentication/authenticators/client/AbstractClientAuthenticator.java#L29). Then select your new authenticator in the admin ui, in the credentials tab of the client settings. The client secret field will disappear.

Screenshot from 2023-08-23 15-10-53
Screenshot from 2023-08-23 15-48-13

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by useresd**
Client Secret field conditional rendering is now depending on the ClientAuthenticatorFactory method supportsSecret(). Which allows the ability to implement the ClientAuthenticator provider that uses a client secret without having the client secret input field disappear from the Admin UI.

Closes #22660

### CODE REVIEW

File: js/apps/admin-ui/src/clients/credentials/Credentials.tsx [hunk 1]
```diff
@@ -200,10 +204,8 @@ export const Credentials = ({ client, save, refresh }: CredentialsProps) => {
               </Button>
             </ActionGroup>
           </CardBody>
-          {(clientAuthenticatorType === "client-secret" ||
-            clientAuthenticatorType === "client-secret-jwt") && <Divider />}
-          {(clientAuthenticatorType === "client-secret" ||
-            clientAuthenticatorType === "client-secret-jwt") && (
+          {selectedProvider?.supportsSecret && <Divider />}
+          {selectedProvider?.supportsSecret && (
```
Comment (jonkoops) [inline js/apps/admin-ui/src/clients/credentials/Credentials.tsx:208]: Looks like we can probably have just one statement here and render both the `Divider` and the `CardBody`, not sure why it wasn't like that in the first place.

**Review by jonkoops — CHANGES_REQUESTED** [review]

File: js/apps/admin-ui/src/clients/credentials/Credentials.tsx (hunk 1 above)
Comment (useresd) [reply js/apps/admin-ui/src/clients/credentials/Credentials.tsx:208]: Okay. I will fix that.

**Review by jonkoops — DISMISSED** [review]
These changes look good to me, I'd still like to see another of the team members take a look at this, especially the Java code.

**Review by ssilvert — DISMISSED** [review]
LGTM

**Comment by jonkoops** [comment]
Hmmm, it looks like these changes are causing some of the tests to fail. Those will have to fixed first before we can merge this change in. I'll do a rebase just to make sure this is not some kind of flake, but I doubt it is.

**Comment by jonkoops** [comment]
Thanks for letting me know @mhajas. I am still curious if the failing Cypress tests have been impacted by this PR, or if the cause was some other work that was merged.

**Comment by useresd** [comment]
The test were failing because of this pull request.

I was able to reproduce the same issues that was on cypress, and it was because of my commit.

The changes I've made to the [buildProviderMetadata](https://github.com/keycloak/keycloak/blob/de448920fcd7ab0e5a4cd17bf7ad6b556c01d6cc/services/src/main/java/org/keycloak/services/resources/admin/AuthenticationManagementResource.java#L164) method affected other endpoints.

I changed the casting of the factory from ConfigurableAuthenticatorFactory to ClientAuthenticatorFactory. But this method is used by other resource that are not an implementation ClientAuthenticatorFactories.

so the following endpoints were affected:
- `/admin/realms/${realmName}/authentication/form-providers`
- `/admin/realms/${realmName}/authentication/authenticator-providers`
- `/admin/realms/${realmName}/authentication/form-action-providers`

I fixed the issue. Did the same failing cypress tests manually and everything seems to work fine now.

I will push the commit.

**Comment by jonkoops** [comment]
Looks like we still have one failing test, but things are already looking better.

**Comment by useresd** [comment]
So, any idea on how to fix that failing test? I can see two tests were failing.

**Comment by jonkoops** [comment]
Looks like in some cases the tests are now getting a `supportsSecret` where they were not before. I don't think that is a bug, so likely we can just update the expected value.

**Comment by jonkoops** [comment]
@ useresd are you still interested in landing this? We have a policy of not keeping PRs out for too long, so let me know if you are still interested.

**Comment by useresd** [comment]
Yes I am. I will change the tests expected value so the tests could pass.

**Comment by ssilvert** [comment]
@useresd Tests still failing.  Please fix.

**Comment by ssilvert** [comment]
@useresd Are you still working on this?

**Comment by useresd** [comment]
Yes. I fixed the test assertion. The test is now passing

```bash
stsuite/integration-arquillian-tests-other/999.0.0-SNAPSHOT/integration-arquillian-tests-other-999.0.0-SNAPSHOT.pom
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for Keycloak Arquillian Integration TestSuite 999.0.0-SNAPSHOT:
[INFO] 
[INFO] Keycloak Arquillian Integration TestSuite .......... SUCCESS [  0.372 s]
[INFO] Test apps .......................................... SUCCESS [  0.008 s]
[INFO] Test apps distribution ............................. SUCCESS [  0.622 s]
[INFO] Keycloak Authz Tests: Hello World Example .......... SUCCESS [  0.460 s]
[INFO] Keycloak Authz: Servlet Authorization Test ......... SUCCESS [  0.231 s]
[INFO] Keycloak Authz: Simple Servlet App with Policy Enforcer SUCCESS [  0.065 s]
[INFO] integration-arquillian-test-apps-servlets .......... SUCCESS [  0.617 s]
[INFO] Keycloak Test App Profile JEE ...................... SUCCESS [  0.090 s]
[INFO] integration-arquillian-test-apps-cors-parent ....... SUCCESS [  0.010 s]
[INFO] Angular Product Portal JS .......................... SUCCESS [  0.259 s]
[INFO] JAX-RS Database Service Using OAuth Bearer Tokens .. SUCCESS [  0.102 s]
[INFO] spring-boot-adapter-app ............................ SUCCESS [  0.203 s]
[INFO] Fuse Test Applications ............................. SUCCESS [  0.008 s]
[INFO] Customer Portal - Secured in Karaf/Fuse ............ SUCCESS [  0.847 s]
[INFO] CXF JAXRS Example - Secured in Karaf/Fuse .......... SUCCESS [  0.201 s]
[INFO] CXF JAXRS Example - Secured in Karaf/Fuse 7 on Undertow SUCCESS [  0.161 s]
[INFO] CXF JAXWS Example - Secured in Karaf/Fuse .......... SUCCESS [  0.117 s]
[INFO] CXF JAXWS Example - Secured in Karaf/Fuse 7 on Undertow SUCCESS [  0.132 s]
[INFO] Product Portal - Secured in Karaf/Fuse ............. SUCCESS [  0.115 s]
[INFO] Product Portal - Secured in Karaf/Fuse 7 on Undertow SUCCESS [  0.150 s]
[INFO] Camel endpoint example - Secured in Karaf/Fuse ..... SUCCESS [  0.143 s]
[INFO] Camel endpoint example - Secured in Karaf/Fuse 7.0 on Undertow SUCCESS [  0.162 s]
[INFO] Keycloak Fuse Example - Features ................... SUCCESS [  0.078 s]
[INFO] Keycloak Examples - External Config ................ SUCCESS [  0.170 s]
[INFO] Servers ............................................ SUCCESS [  0.010 s]
[INFO] Testsuite Adapter SPI .............................. SUCCESS [  0.007 s]
[INFO] Adapter Undertow SPI (JakartaEE) ................... SUCCESS [  0.355 s]
[INFO] Undertow OIDC Adapter (JakartaEE) .................. SUCCESS [  0.423 s]
[INFO] Undertow SAML Adapter (JakartaEE) .................. SUCCESS [  0.438 s]
[INFO] Auth Server ........................................ SUCCESS [  0.008 s]
[INFO] Auth Server Services ............................... SUCCESS [  0.010 s]
[INFO] Auth Server Services - Testsuite Providers Deployment SUCCESS [  0.060 s]
[INFO] Auth Server Services - Testsuite Providers ......... SUCCESS [  1.381 s]
[INFO] Test Util .......................................... SUCCESS [  0.470 s]
[INFO] Auth Server - Undertow ............................. SUCCESS [  0.640 s]
[INFO] Auth Server - Quarkus .............................. SUCCESS [  8.211 s]
[INFO] App Server ......................................... SUCCESS [  0.005 s]
[INFO] App Server - SPI ................................... SUCCESS [  0.043 s]
[INFO] App Server - JBoss ................................. SUCCESS [  0.007 s]
[INFO] App Server - Wildfly ............................... SUCCESS [ 25.993 s]
[INFO] App Server - Karaf ................................. SUCCESS [  0.005 s]
[INFO] App Server - Tomcat ................................ SUCCESS [  0.006 s]
[INFO] App Server - Undertow .............................. SUCCESS [  0.257 s]
[INFO] App Server - Jetty Parent .......................... SUCCESS [  0.009 s]
[INFO] Keycloak Arquillian Integration TestSuite - Cache Server SUCCESS [  0.008 s]
[INFO] Keycloak Arquillian Integration TestSuite - Cache Server - Infinispan SUCCESS [  0.009 s]
[INFO] Keycloak Arquillian Integration TestSuite - Cache Server - Legacy SUCCESS [  0.009 s]
[INFO] Tests .............................................. SUCCESS [  0.058 s]
[INFO] Base TestSuite ..................................... SUCCESS [ 28.615 s]
[INFO] Other Tests Modules ................................ SUCCESS [  0.006 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:13 min
[INFO] Finished at: 2023-09-27T14:42:15+03:00
[INFO] ------------------------------------------------------------------------

```

**Comment by jonkoops** [comment]
@useresd there are still failing tests on CI, and it seems that the branch now has conflicts with `main`. Could you resolve them and take a look at fixing the tests?

**Comment by jonkoops** [comment]
@useresd are you still interested in landing this PR? If so let us know, otherwise we will close it to prune out our backlog.

**Comment by useresd** [comment]
> @useresd are you still interested in landing this PR? If so let us know, otherwise we will close it to prune out our backlog.

Sure. I am interested in landing this PR. I will fix the conflict and make sure the other tests are passing.

**Review by jonkoops — APPROVED** [review]
Front-end changes look good to me, @ssilvert could you take a look at the other stuff?

**Review by ghost — COMMENTED** [review]
Unreported flaky test detected, please review

**Review by ssilvert — APPROVED** [review]
LGTM

**Comment by ssilvert** [comment]
@useresd Please fix conflicts.

**Comment by jonkoops** [comment]
@useresd Are you still interested in landing this? We have a convention of closing PRs that have been out for too long without activity, so we'll likely be closing this one soon.

**Comment by ImFlog** [comment]
I have the exact same issue and I am interested in the PR.
If @useresd don't have enough time to help on this I could restart the MR with the same code and fix conflicts.
WDYT ?

**Comment by useresd** [comment]
I tried to fix failing tests, but more tests keep failing. Running the tests locally on my machine takes a lot of times and consumes bandwidth. I will try to give it a final shot today.

**Comment by jonkoops** [comment]
Closing this in favor of #25776.

**Pull request closed without merging**